apiserver/common/instanceidgetter.go

```diff
@@ -59,15 +49,27 @@ func (ig *InstanceIdGetter) InstanceId(ctx context.Context, args params.Entities
 			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
 			continue
 		}
-		err = apiservererrors.ErrPerm
 		if canRead(tag) {
-			var instanceId instance.Id
-			instanceId, err = ig.getInstanceId(tag)
-			if err == nil {
-				result.Results[i].Result = string(instanceId)
+			machineUUID, err := ig.machineService.GetMachineUUID(ctx, machine.Name(tag.Id()))
+			if errors.Is(err, machineerrors.MachineNotFound) {
+				result.Results[i].Error = apiservererrors.ServerError(errors.NotFoundf("machine %s", tag.Id()))
+				continue
+			} else if err != nil {
+				result.Results[i].Error = apiservererrors.ServerError(err)
+				continue
+			}
+			instanceId, err := ig.machineService.InstanceID(ctx, machineUUID)
+			if errors.Is(err, machineerrors.NotProvisioned) {
+				result.Results[i].Error = apiservererrors.ServerError(errors.NotProvisionedf("machine %s", tag.Id()))
+				continue
+			} else if err != nil {
+				result.Results[i].Error = apiservererrors.ServerError(err)
+				continue
 			}
+			result.Results[i].Result = instanceId
+		} else {
+			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
```

Comment:
This is messy, can we move it to a method `getInstanceIdForOne` or something similar?

Reply:
I was going to do that, but passing back and forth the errors and results seemed worst than this, furthermore the function is not that long (~30 lines).